src/libexpr/primops/filterPath.cc

```diff
@@ -0,0 +1,118 @@
+
+    state.forceAttrs(*args[0], pos,
+        "while evaluating the first argument to 'builtins.filterPath'");
+
+    for (auto & attr : *args[0]->attrs) {
+        auto n = state.symbols[attr.name];
+        if (n == "path")
+            path.emplace(state.coerceToPath(attr.pos, *attr.value, context,
+                    "while evaluating the 'path' attribute passed to 'builtins.filterPath'"));
+        else if (n == "filter") {
+            state.forceValue(*attr.value, pos);
+            filterFun = attr.value;
+        }
+        else
+            state.debugThrowLastTrace(EvalError({
+                .msg = hintfmt("unsupported argument '%1%' to 'filterPath'", state.symbols[attr.name]),
+                .errPos = state.positions[attr.pos]
+            }));
+    }
+
+    if (!path)
+        state.debugThrowLastTrace(EvalError({
+            .msg = hintfmt("'path' required"),
+            .errPos = state.positions[pos]
+        }));
+
+    if (!filterFun)
+        state.debugThrowLastTrace(EvalError({
+            .msg = hintfmt("'filter' required"),
+            .errPos = state.positions[pos]
+        }));
+
+    if (!context.empty())
+        state.debugThrowLastTrace(EvalError({
+            .msg = hintfmt("'path' argument to 'filterPath' cannot have a context"),
+            .errPos = state.positions[pos]
+        }));
+
+    auto accessor = make_ref<FilterPathInputAccessor>(state, pos, *path, filterFun);
+
+    state.registerAccessor(accessor);
+
+    v.mkPath(accessor->root());
+}
+
+static RegisterPrimOp primop_filterPath({
+    .name = "__filterPath",
+    .args = {"args"},
+    .doc = R"(
+      This function lets you filter out files from a path. It takes a
+      path and a predicate function, and returns a new path from which
+      every file has been removed for which the predicate function
+      returns `false`.
+
+      For example, the following filters out all regular files in
+      `./doc` that don't end with the extension `.md`:
+
+      ```nix
+      builtins.filterPath {
+        path = ./doc;
+        filter =
+          path: type:
+          (type != "regular" || hasSuffix ".md" path);
+      }
+      ```
+
+      The filter function is called for all files in `path`. It takes
+      two arguments. The first is a string that represents the path of
+      the file to be filtered, relative to `path` (i.e. it does *not*
+      contain `./doc` in the example above). The second is the file
+      type, which can be one of `regular`, `directory` or `symlink`.
+
+      Note that unlike `builtins.filterSource` and `builtins.path`,
+      this function does not copy the result to the Nix store. Rather,
+      the result is a virtual path that lazily applies the filter
+      predicate. The result will only be copied to the Nix store if
+      needed (e.g. if used in a derivation attribute like `src =
+      builtins.filterPath { ... }`).
+    )",
+    .fun = prim_filterPath,
+    .experimentalFeature = Xp::Flakes,
+});
```

Comment:
`filterPath` as implemented allows a new species of path to be constructed; a lot like a file, but with an odd `dirOf`.

```
nix-repl> p = builtins.filterPath { path = ./default.nix; filter = _: _: true; }

nix-repl> dirOf p == p
true

nix-repl> p2 = ./default.nix

nix-repl> dirOf p2 == p2
false

```

This seems undesirable, especially while comparing `dirOf` is the only way to find the tree root, but also if that was provided for.

The "outPath" behavior is also different.

```
nix-repl> "${./default.nix}"
"/nix/store/i09iwz8gh5qcrhmgfiz8g381b5qky4v1-default.nix"

nix-repl> "${p}"             
"/nix/store/n3afs2np12qjb7lgj8n54rpjs2hzar8j-source"

```

While these design choices are by no means easy to compare, I think the least surprising choice is the one where the complexity can be understood locally - even when relevant events occur in code written by different people.
"filterPath returns a file path instead of a tree when the path is a file" seems like a rule that's easier to understand and troubleshoot than "a value that behaves like a path may behave differently in rare cases", as a user would have no reason to suspect that this might happen, and there's no ad hoc repl test that explains the behavior.
So I think `filterPath` should make an exception for a `path` argument. When it's a single file (or symlink I suppose), return that as a regular path instead of a lazy tree.

Reply:
Alternatively, could returning the file as part of a lazy directory with just that file in it work? Returning a regular path would still cause the file to be copied to the store, which might still be an issue especially in case of large files, right?

Reply:
I think you're right. Returning a single directory with just the file is better than returning the file in its original root, as it would be reasonable to assume that no path outside the `path` argument can be recovered from the returned path.

> Returning a regular path would still cause the file to be copied to the store

That wouldn't be necessary until used as a string.

Reply:
`filterPath` is not referentially transparent under `==` equality.

```
nix-repl> d1 = builtins.filterPath { path = ./.; filter = always; }     

nix-repl> d2 = builtins.filterPath { path = ./.; filter = always; } 

nix-repl> d1 == d2
false  # wrong
```

The default equality should be implemented by traversing the tree and making sure that the trees behave equally. That is still quite lazy, as no copying is necessary. It should compare the directory structure and outcome of the `filter` applications need to be computed, and if that all works out, it should check that either the files are in the same physical location, or have equal contents. The latter would rarely be necessary.